Address AUTH key handling and epoch handling

tuexen opened the pull request:
This should describe the handling of AUTH keys such that we are sure that we are not loosing any messages due to missing AUTH keys. It also describes the epoch handling, but this still has some issues.

emanjon commented:
Comments on the SCTP-AUTH part


>   Before sending the DTLS Finished message, the active SCTP-AUTH key
>   SHOULD be switched to the new one, it MAY be sent using the old one
>   (for example in the case of the client performing a full handshake or the
>   server performing a session-resuming handshake).
>   If the active SCTP-AUTH key is not switched to the new one before sending
>   the DTLS Finished message, it MUST be switched to the new one immediately
>   after sending the DTLS Finished message.

On the client side, the Finished message is likely send in the same DTLS record as the
ChangeCipherSpec. While the TLS implementation has all information to derive the
master secret after generating ChangeCipherSpec, I don't know when the new master
key is made available to the exporter, I think this is unspecified....

Maybe it is best to have "After sending the DTLS Finished message," for both
DTLS 1.2 and DTLS 1.3

> (for example in the case of the client performing a full handshake or the
> server performing a session-resuming handshake).

This list seems exhaustive.

>   Once the Finished message using DTLS epoch n with n > 2 has been processed
>   by DTLS, the SCTP-AUTH key with Shared Key Identifier n - 2 MUST be removed.

I think this should be "n > 3" as 0 has already been removed.

emanjon commented:
Comments on the DTLS parts:

>Before sending during renegotiation a ClientHello message or ServerHello
>   message, 

The server send a HelloRequest message to initiate renegotiation (Not ServerHello).
Could we say just: "Before inititating or accepting renegotiation"?



>User messages that arrive between ChangeCipherSpec and Finished messages and
>   use the new epoch have probably passed the Finished message and MUST be
>   buffered by DTLS until the Finished message has been processed.

Should this be user messages? user messages might be processed with several different epochs I assume.
Would "user messages where SCTP chunks arrive between..." be better?

emanjon commented:
I assume updating key after ChangeCipherSpec is standards. But what about the reciving side? I think we need to have a requirement here that ChangeCipherSpec is send using old SCTP-AUTH key, otherwise the receiving will reject ChangeCipherSpec before processing it. We probably don't want to mandate anything for the DTLS  layer.

tuexen commented:
> Comments on the SCTP-AUTH part
> 
> > Before sending the DTLS Finished message, the active SCTP-AUTH key
> > SHOULD be switched to the new one, it MAY be sent using the old one
> > (for example in the case of the client performing a full handshake or the
> > server performing a session-resuming handshake).
> > If the active SCTP-AUTH key is not switched to the new one before sending
> > the DTLS Finished message, it MUST be switched to the new one immediately
> > after sending the DTLS Finished message.
> 
> On the client side, the Finished message is likely send in the same DTLS record as the
> ChangeCipherSpec. While the TLS implementation has all information to derive the
> master secret after generating ChangeCipherSpec, I don't know when the new master
> key is made available to the exporter, I think this is unspecified....

I think the Finished and the ChangeCipherSpec messages each have its own DTLS record header,
but I think both records are often put in the same UDP packet.

I agree, that it is left unspecified when the exporter (using some API) can be used.
Since implementing SCTP/DTLS has to touch code in the DTLS implementation, I was
assuming that this could be changed to fit our needs...

> 
> Maybe it is best to have "After sending the DTLS Finished message," for both
> DTLS 1.2 and DTLS 1.3

I'm fine with that.

> 
> > (for example in the case of the client performing a full handshake or the
> > server performing a session-resuming handshake).
> 
> This list seems exhaustive.

It is not intended to cover the server side performing a full handshake or
the client side performing a session-resuming handshake. In both cases
the peer can make sure the SCTP AUTH keys are configured when the
corresponding packets are received.

> 
> > Once the Finished message using DTLS epoch n with n > 2 has been processed
> > by DTLS, the SCTP-AUTH key with Shared Key Identifier n - 2 MUST be removed.
> 
> I think this should be "n > 3" as 0 has already been removed.

I think the text is correct. The smallest applicable number is 3 and this would result
in removing 3 - 2 = 1, which is not 0.

tuexen commented:
> Comments on the DTLS parts:
> 
> > Before sending during renegotiation a ClientHello message or ServerHello
> > message,
> 
> The server send a HelloRequest message to initiate renegotiation (Not ServerHello).
> Could we say just: "Before inititating or accepting renegotiation"?

Sure. It would be OK, if the client does this check before sending the ClientHello message in response to the HelloRequest, but we can change roles here. I don't think it makes a difference.

But then the next paragraph becomes more complex. 

I can change both paragraphs if you want. Just let me know.

> 
> > User messages that arrive between ChangeCipherSpec and Finished messages and
> > use the new epoch have probably passed the Finished message and MUST be
> > buffered by DTLS until the Finished message has been processed.
> 
> Should this be user messages? user messages might be processed with several different epochs I assume.
> Would "user messages where SCTP chunks arrive between..." be better?

But is it OK to process user message of a new epoch *before* processing the Finished message? Couldn't this be used by an attacker to get message delivered to the application and bypassing the checks done by processing the Finished message?

tuexen commented:
> I assume updating key after ChangeCipherSpec is standards. But what about the reciving side? I think we need to have a requirement here that ChangeCipherSpec is send using old SCTP-AUTH key, otherwise the receiving will reject ChangeCipherSpec before processing it. We probably don't want to mandate anything for the DTLS layer.

I agree with both points. Were is the text contradicting this?

emanjon commented:
"previous epoch" could maybe be misunderstood...  Yes, I think it should be something like:

   Before inititating or accepting renegotiation resulting in epoch n+1, the DTLS endpoint MUST ensure
   that all DTLS messages using epoch n-1 have been acknowledged by
   the SCTP peer in a non-revokable way.

I don't think the endpoints can change roles. 



I think the following is handled by DTLS:

   User messages that arrive between ChangeCipherSpec and Finished messages and
   use the new epoch have probably passed the Finished message and MUST be
   buffered by DTLS until the Finished message has been processed.


I don't see why the following is needed as the requires as the endpoint has both the new and the old key in parallel.

   Prior to processing a received ClientHello message or ServerHello
   message, all other received SCTP user messages that are buffered in the
   SCTP layer and can be delivered to the DTLS layer MUST be read and
   processed by DTLS.

tuexen commented:
> "previous epoch" could maybe be misunderstood... Yes, I think it should be something like:
> 
> Before inititating or accepting renegotiation resulting in epoch n+1, the DTLS endpoint MUST ensure
> that all DTLS messages using epoch n-1 have been acknowledged by
> the SCTP peer in a non-revokable way.
> 
> I don't think the endpoints can change roles.

No, they can't. But using the text I suggested, the client drains first, then the server, no matter which side triggered the renegotiation.

In your text, if the client triggers the renegotiation, the client drains first, then the server. If the server triggers the renegotiation, the server drains first, then the client drains.
I'm not sure if you consider this different important or not.

> 
> I think the following is handled by DTLS:
> 
> User messages that arrive between ChangeCipherSpec and Finished messages and
> use the new epoch have probably passed the Finished message and MUST be
> buffered by DTLS until the Finished message has been processed.

Yes, you are right. This is handled in DTLS 1.2, but wasn't in DTLS 1.0. That is why we added it to RFC 6083.
I've removed it.
> 
> I don't see why the following is needed as the requires as the endpoint has both the new and the old key in parallel.
> 
> Prior to processing a received ClientHello message or ServerHello
> message, all other received SCTP user messages that are buffered in the
> SCTP layer and can be delivered to the DTLS layer MUST be read and
> processed by DTLS.

As far as I understood, draining and waiting is no an option. So the idea was to have always (when it makes sense) the current and the key before active to avoid message loss due to missing key material.

Here is the procedure:

Looking at data sent from the client to the server. The client uses epoch n. Before sending the ClientHello message, the client must stop and wait until all data using epoch n - 1 has the acknowledged by the peer. That means, when the server reads the ClientHello, all data using epoch n - 1 which has not been processed by the DTLS layer is at least buffered at the SCTP layer. Now reading and processing it ensures that all data using epoch n - 1 has been completely processed before switching to epoch n + 1. This means the DTLS layer can drop the keys for epoch n - 1 when switching to epoch n + 1. So for receiving the DTLS layer only needs the key material of two epochs, not more.

The same applies for data sent from the server to the client.

Does this makes sense? Is it acceptable?

emanjon commented:
>After sending the KeyUpdate message, it stops sending DTLS messages until
  >the corresponding Ack message has been processed.
  
  Is this needed? The recipient could just buffer the DTLS records until the KeyUpdate has been received.

tuexen commented:
> > After sending the KeyUpdate message, it stops sending DTLS messages until
> > the corresponding Ack message has been processed.
> 
> Is this needed? The recipient could just buffer the DTLS records until the KeyUpdate has been received.

How do you make sure that all messages using an old epoch n - i have been processed before you remove the keys for epoch n - i?

gloinul commented on draft-westerlund-tsvwg-dtls-over-sctp-bis.md:
```diff
@@ -508,40 +508,93 @@ ULP:  Upper Layer Protocol
+
+   The procedures of Section 4.2.1 of {{RFC6347}} are irrelevant.
+   When receiving DTLS packets using epoch n, no DTLS packets from earlier
+   epochs are received.
```
This implies draining of protection records. Couldn't we avoid this.

tuexen replied:
Right now, I don't know now.

gloinul commented:
I think we should fix the easy errors and then merge and create a new issue regarding DTLS 1.3 records handling in user message over key updates.

tuexen commented:
> I think we should fix the easy errors and then merge and create a new issue regarding DTLS 1.3 records handling in user message over key updates.

I agree.

gloinul commented:
I am fixing the reference and the roll over. Will commit soon.

gloinul reviewed (approved):
I think we can merge this, but will create a new issue that is more specific to the remaining aspect of DTLS epoch handling.

tuexen commented on draft-westerlund-tsvwg-dtls-over-sctp-bis.md:
```diff
@@ -508,40 +508,93 @@ ULP:  Upper Layer Protocol
+   Before sending during renegotiation a ClientHello message or ServerHello
+   message, the DTLS endpoint MUST ensure that all DTLS messages using the
+   previous epoch have been acknowledged by the SCTP peer in a non-revokable
+   way.
```
For me a DTLS message consists of a DTLS record. In case of UDP. In case of UDP, you can put multiple DTLS records or DTLS messages in a single UDP packet. In case of SCTP you can put multiple DTLS messages in a single SCTP user message. This is what this document is about. RFC 6083 has a one-to-one relation between DTLS messages / DTLS records and SCTP user messages.

Are you asking for a wording like `MUST ensure that all DATA chunks related to DTLS messages using the previous epoch have been`?

tuexen commented on draft-westerlund-tsvwg-dtls-over-sctp-bis.md:
```diff
@@ -508,40 +508,93 @@ ULP:  Upper Layer Protocol
+   The exporter MUST use the label given in Section {{IANA-Consideration}}
+   and no context.
+   The new Shared Key Identifier MUST be the old Shared Key Identifier
+   incremented by 1.
```
As far as I know the epoch can't wrap in DTLS 1.2 and DTLS 1.3 anymore. Since the epoch and the key id are both 16 bit, the key id can't wrap anymore. In DTLS 1.0 the epoch could wrap, therefore the key id.
